### ISSUE TITLE
Column multi-index does not succeed roundtrip

### ISSUE THREAD

**Issue by yohplala** [dask/fastparquet#728]
**What happened**:
Reading the code managing column multi-index, there is a line I don't understand.
Trying on a test case, it shows a bug, but not completely sure the bug raised is indeed because of this line, or because of something else.

So, first, line I don't understand is [line 747](https://github.com/dask/fastparquet/blob/9bd2a57a627f08629e4b97d22df4f4ccc4c3e0ce/fastparquet/writer.py#L747) in `writer.make_metadata`.
```python
    if isinstance(data.columns, pd.MultiIndex):
        name = data.index.name or "index"                # this line
        index_cols = [{'field_name': name,
                       'metadata': None,
                       'name': name,
                       'numpy_type': 'object',
                       'pandas_type': 'mixed-integer'}]
        ci = [
            get_column_metadata(ser, n)
            for ser, n
            in zip(data.columns.levels, data.columns.names)
        ]
```
My understanding is that in pandas, `df.columns` is column index, and `df.index` is row index.
I understand from this line that the name stored for the column index is that of the row index. Is this ok?

Regarding now the failing example, I don't see how much the pandas dataframe differs from the test case already existing, but here it is, as well as the error message.
```python
import os
import pandas as pd
import fastparquet as fp
import numpy as np

tempdir = os.path.expanduser('~/Documents/code/data/fastparquet/cmidx')

ts = [pd.Timestamp('2021/01/01 08:00:00'),
      pd.Timestamp('2021/01/05 10:00:00')]
val = [10, 34]

df = pd.DataFrame({'val': val, 'ts': ts})
tuples = [(col, point) for col, point in zip(df.columns, ['0', ''])]
cmidx = pd.MultiIndex.from_tuples(tuples, names=('component', 'point'))
df.columns = cmidx

fp.write(tempdir, df, file_scheme='hive')
df_out = fp.ParquetFile(tempdir).to_pandas()
```
Error message:
```python
ValueError: Following columns were requested but are not available: {'index'}.
All requested columns: ["('val', '0')", 'ts', 'index']
Available columns: ["('val', '0')", 'ts']
```
**Environment**:

- Fastparquet version: from PR #712 (ok, ok, I am not doing the safest thing ever here :))
- Python version: 3.9.7
- Operating System: Ubunut 20.04
- Install method: `pip install -e .`

**Comment by yohplala** [comment]
Because of the error message, it seems to me that the emtpy 2nd level is not kept, and this may be another trouble.
Initial cmidx (column multi-index) is
```python
df.columns
Out[2]: 
MultiIndex([('val', '0'),
            ( 'ts',  '')],
           names=['component', 'point'])
```
But error message shows the empty string is not kept.

**Comment by martindurant** [comment]
I have no answer: obviously the error should not happen. Indeed, the name of the index should have no bearing on what is stored for the columns.

> it seems to me that the empty 2nd level is not kept

I *think* this is fine, although in the tests we either have simple columns names, or all levels defined.

**Comment by yohplala** [comment]
> I _think_ this is fine, although in the tests we either have simple columns names, or all levels defined.

For information, I am investigating.
From these 2 tests, I am concluding we should keep the empty string and a tuple, or at least a string in a tuple.
I am checking how doing so.

```python
dft = pd.DataFrame({('a','b'):[1,2,3],'c':[4,5,6]})

dft
Out[56]: 
   (a, b)  c
0       1  4
1       2  5
2       3  6

dft.columns
Out[57]: Index([('a', 'b'), 'c'], dtype='object')  # is understood by pandas as a standard index
```
```python
dft = pd.DataFrame({('a','b'):[1,2,3],('c','d'):[4,5,6]})

dft
Out[59]: 
   a  c
   b  d
0  1  4
1  2  5
2  3  6

dft.columns
Out[60]:            # is understood by pandas as a multi-index
MultiIndex([('a', 'b'),
            ('c', 'd')],
           )
```

**Comment by martindurant** [comment]
Doesn't fastparquet explicitly make a MultiIndex, though?

**Comment by yohplala** [comment]
> Doesn't fastparquet explicitly make a MultiIndex, though?

Before reaching this step, there is a trouble in `to_pandas()` -> `pre_allocate()` -> `ast.eval_literal()` through which goes the columns names as tuples. `ast.eval_literal()` raises an error with a simple string that does not embed a tuple.

In front of 2 paths:
 - reworking the code at the reading step here to make it accept a string not being a tuple
 - or reworking the code when the recording is done so that the column name is correctly recorded as a tuple

I have chosen the 2nd because of how pandas behaves in the 2 tests above.
After ...1 hour should I say... of looking for where is done this change from a tuple to a single string, I have found it!
In `utils.py` -> `norm_col_name()`
Removing 'faulty' lines, this works... but some test cases are not passing anylonger.
```python
def norm_col_name(name):
    if isinstance(name, tuple):
#R
#        name = tuple(filter(bool, name))
#        if len(name) == 1:
#            return name[0]
        return str(name)
    return name
```
I can check tomorrow, I don't know if it is a 'real' problem or not.
But I can also check the 1st track I have ignored: reworking the way fastparquet reads column names, i.e enclose `ast.literal_eval` within a `try/except` maybe?

Would you have an advice for that?
Thanks in advance for your feedback,
Bests,

**Comment by yohplala** [comment]
@martindurant 
I implemented a fix in https://github.com/yohplala/fastparquet/tree/column_multi_index
I pushed it to Github only to have a copy, but it is full of `print` statements, not clean at all. (also, it is starting from branch `write_row_groups`)
I will start a new branch from current `main`, clean, add above example as test case and PR probably beginning of next week.

I ended up keeping empty strings in tuples for 'normal columns', because I have been thinking to the case (possibly not very likely) when a user will name his/her column `('','something')` i.e. starting with an empty string on 1st level, then something on 2nd level. Or something even more awful:
`('','something','','just','','for','','kidding','')`   :)

BUT, when index are reset into column, for this type of column, I kept applying `norm_col_name`, because in this case, we know sub levels will always be empty, and resetting is an action from fastparquet.


PS: my god, such a dataframe really is possible:
```python
df = pd.DataFrame({('','a','','b',''): [1,2],
                   ('o','','i','','k'):[3,4]})

df
Out[57]: 
      o
   a   
      i
   b   
      k
0  1  3
1  2  4

df.columns
Out[58]: 
MultiIndex([( '', 'a',  '', 'b',  ''),
            ('o',  '', 'i',  '', 'k')],
           )
```
